added LeadershipAbdicationIntegrationTest and change deployment handling during leader abdication

unterstein opened the pull request:
`LeadershipAbdicationIntegrationTest` was taken from https://github.com/mesosphere/marathon/pull/4623 and adapted to current test classes.

Furthermore the claimed behavior was corrected to `keep running tasks after abdication` through the change in `AppStartActor`.

Fixes #3783

review-ninja commented:
<a href="https://app.review.ninja/mesosphere/marathon/pull/4900" target="_blank"><img src="https://app.review.ninja/10941409/pull/4900/badge" alt="ReviewNinja"/>

aquamatthias commented on src/test/scala/mesosphere/marathon/integration/LeadershipAbdicationIntegrationTest.scala:
```diff
@@ -0,0 +1,90 @@
+    tasksBeforeAbdication should be (tasksAfterFirstAbdication)
+
+    When("the leader abdicates")
+    val secondResult = firstProcess.client.abdicate()
```
Looks not very DRY - can we move the abdication logic into a function?

unterstein replied:
Did a loop like, Karsten suggested.

aquamatthias commented on src/test/scala/mesosphere/marathon/integration/LeadershipAbdicationIntegrationTest.scala:
```diff
@@ -0,0 +1,90 @@
+
+  after(cleanUp())
+
+  override val numAdditionalMarathons = 5
```
A) this number is pretty high
B) the number of abdications is the number of additional marathons - right?

Can we write this test in a way, that this number controls both.
Perhaps we can lower this number?

unterstein replied:
Sure

unterstein commented:
Please don`t merge yet. I am currently addressing feedback, rebasing and changing commit history to have:
1.) Fix from aleks
2.) IT from jparkie

Please do `rebase and merge` afterwards

jeschkies commented on src/main/scala/mesosphere/marathon/upgrade/AppStartActor.scala:
```diff
@@ -23,14 +25,17 @@ class AppStartActor(
-  val nrToStart: Int = scaleTo
+  override val nrToStart: Int = scaleTo
 
   def initializeStart(): Unit = {
```
Should `initializeStart` be asynchrone?

zen-dog replied:
If you want to avoid using `specInstancesSync`

jeschkies commented on src/test/scala/mesosphere/marathon/integration/LeadershipAbdicationIntegrationTest.scala:
```diff
@@ -0,0 +1,90 @@
+  override val numAdditionalMarathons = 5
+
+  def firstProcess = runningServerProcesses.headOption.getOrElse(
+    fail("there are marathon servers running")
```
I think the error message is wrong. It should say "no servers are running".

jeschkies commented on src/test/scala/mesosphere/marathon/integration/LeadershipAbdicationIntegrationTest.scala:
```diff
@@ -0,0 +1,90 @@
+
+  test("Abdicating a leader does not kill a running task") {
+    Given("a new app with an impossible constraint")
+    // Running locally, the constraint of a unique hostname should prevent the second instance from deploying.
```
What's the motivation for this?

jeschkies commented on src/test/scala/mesosphere/marathon/integration/LeadershipAbdicationIntegrationTest.scala:
```diff
@@ -0,0 +1,90 @@
+    Then("the task should not be killed")
+    tasksBeforeAbdication should be (tasksAfterSecondAbdication)
+
+    When("the leader abdicates")
```
Can we do this in a loop?

unterstein replied:
Yes

unterstein commented:
Thanks for the feedback @aquamatthias, @jeschkies. Here is the background:
Yelp witnessed the behavior that during an abdication while running deployments leads to killing of all tasks, see #3783 and https://github.com/EvanKrall/reproduce_marathon_issue_3783. They made the PR #4623 provide a test to reproduce this, which is the first commit of this PR. The idea of this PR was to keep the original author of the Test and fix the actual problem of Aleksey.

I added a third commit, addressed your feedback, made it more clear and added documentation.
